### PR TITLE
Json & XML Dbcodecs

### PR DESCRIPTION
**Pull request by AugustNagro**
Resolves https://github.com/AugustNagro/magnum/issues/27

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -529,6 +529,86 @@ If instead your Postgres type is an array of varchar or text, use the following
...
+```scala
+import com.augustnagro.magnum.pg.json.JsonDbCodec
+import play.api.libs.json.*
+
+trait PlayJsonDbCodec[A] extends JsonDbCodec[A]
+
+object PlayJsonDbCodec:
+  def derived[A](using jsonCodec: OFormat[A]): PlayJsonDbCodec[A] = new:
+    def encode(a: A): String = jsonCodec.writes(a).toString
+    def decode(json: String): A = jsonCodec.reads(Json.parse(json)).get
```
Comment (guizmaii) [inline README.md:583]: I wonder if we can't provide these instances thanks to the `Optional` magic of sbt 🤔
See https://blog.7mind.io/no-more-orphans

File: magnum-pg/src/main/scala/com/augustnagro/magnum/pg/json/JsonBDbCodec.scala [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+
+trait JsonBDbCodec[A] extends DbCodec[A]:
+
+  def encode(a: A): String
+
+  def decode(json: String): A
+
+  override def queryRepr: String = "?"
+
+  override val cols: IArray[Int] = IArray(Types.OTHER)
```
Comment (guizmaii) [inline magnum-pg/src/main/scala/com/augustnagro/magnum/pg/json/JsonBDbCodec.scala:16]: Why `Types.OTHER` here? 🤔
I implemented this way: `IArray(Types.JAVA_OBJECT)`. Not sure to remember why. Probably something I found on internet

File: magnum-pg/src/main/scala/com/augustnagro/magnum/pg/json/JsonDbCodec.scala [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+
+  def decode(json: String): A
+
+  override def queryRepr: String = "?"
+
+  override val cols: IArray[Int] = IArray(Types.OTHER)
+
+  override def readSingle(resultSet: ResultSet, pos: Int): A =
+    val rawJson = resultSet.getString(pos)
+    if rawJson == null then null.asInstanceOf[A]
```
Comment (guizmaii) [inline magnum-pg/src/main/scala/com/augustnagro/magnum/pg/json/JsonDbCodec.scala:20]:
```suggestion
    if rawJson eq null then null.asInstanceOf[A]
```
is probably enough

**Review by guizmaii — COMMENTED** [review]
Made a comment you're free to ignore as it can be more complicated to implement: https://github.com/AugustNagro/magnum/pull/49/files#r1807555435

Otherwise, I like the approach :)

File: README.md (hunk 1 above)
Comment (AugustNagro) [reply README.md:583]: Wow.. that's mind blowing.. thanks for sharing, I'll try it out.

File: magnum-pg/src/main/scala/com/augustnagro/magnum/pg/json/JsonBDbCodec.scala (hunk 2 above)
Comment (AugustNagro) [reply magnum-pg/src/main/scala/com/augustnagro/magnum/pg/json/JsonBDbCodec.scala:16]: The PgCodecTests "insert MagCar" test fails when using JAVA_OBJECT but works with OTHER

```
INSERT INTO mag_car (id, text_colors, text_color_map, last_service, my_json_b, my_xml) VALUES (?, ?, ?, ?, ?, ?)
With message:
org.postgresql.util.PSQLException: Unknown Types value.
```

File: magnum-pg/src/main/scala/com/augustnagro/magnum/pg/json/JsonDbCodec.scala (hunk 3 above)
Comment (AugustNagro) [reply magnum-pg/src/main/scala/com/augustnagro/magnum/pg/json/JsonDbCodec.scala]: Done :)

**Comment by AugustNagro** [comment]
Thanks for the review @guizmaii . It's a very interesting suggestion to use https://blog.7mind.io/no-more-orphans , I like the idea a lot.

I wonder can we merge these codecs as-is, and then experiment with adding the Optional codecs per https://blog.7mind.io/no-more-orphans in a new MR? I want to make sure I fully understand the implications.

File: README.md (hunk 1 above)
Comment (AugustNagro) [reply README.md]: Filed https://github.com/AugustNagro/magnum/issues/67

I believe we can merge this as-is and add these instances in a new ticket.

**Pull request merged**